RFC: Improved suggestions for namespace use

kilroyd opened the pull request:
This series is looking to address #828, where IWYU just suggests the first header that declares a namespace to satisfy a using directive.

The first couple commits play with namespace aliases, which were easier to get a handle on. The first commit adds reporting for when namespace aliases use a namespace, and the second adds a warning when the alias is unused. Although useful, I suspect we want to drop the second commmit (because IWYU generally doesn't warn on unused declarations). I believe the first commit is technically correct, but I've not seen a codebase with aliases in headers so hopefully this won't be disruptive.

The third commit finally attempts to improve the suggestion for the inculde to use. The logic is mostly in a new file, iwyu_namespace.cc, as I was trying a bunch of different things. I ended up caching the ReportDeclUse() to call during ResolvePendingAnalysis(), so we can pick the prefered NamespaceDecl to report. This works for the simple test case.

Any feedback appreciated!

kimgr commented:
Wow, what an ambitious patch! It's going to take me a while to chew through, but it looks really nice and clean at first browse-through.

kilroyd commented:
Pushed an updated series:

* fixing issues reported by CI.
* improve test to use nested namespaces
* handle nested namespaces

bolshakov-a commented:
I haven't taken closer look at your changes, but I've just recalled that I also have a draft commit addressing the namespace issue, FYU:
https://github.com/bolshakov-a/include-what-you-use/commit/e1d6996ea187d1350ff1e5bf2079b0be0563f3f7

kimgr commented:
Sorry, I've been swamped with life in the past weeks, but I'm still hoping to get to reviewing this soon. My gut instinct says it shouldn't be it's own module (`iwyu_namespace.*`), but I see how the additional data structures would probably take up a lot of conceptual space in the existing modules. Let me see if I can get my head around it and think about alternatives. (@bolshakov-a's patch looks much smaller, so it would be interesting to hear if they solve different problems)

kilroyd commented:
I just had a look at @bolshakov-a's commit. I had to think a bit about what it's doing. If I understood correctly:

* if we have a namespace-use (i.e. namespace directive) ignore it if we find any other identifier use from a sub-namespace. We can do this on the basis that the other identifier use will necessarily require an appropriate header inclusion. My commit effectively resorts to this in the 'we have many namespace decls with declarations in them' case. This also works for the namespace alias support as added by my first commit in this PR.

My commit is _much_ heavier handed (tracking how many decls in each `NamespaceDecl` we see) - but helps in the case where there is no other identifier use ++. It can still suggest an appropriate header. One other affect of my commit is that it will still note that we get the namespace from a particular header in the include comments.

For the common case, I think Andrey's commit is good. We could commit both to get a good suggestion for the namespace when it's otherwise unused, though I'm not sure the extra logic is worth it.

> My gut instinct says it shouldn't be it's own module (iwyu_namespace.*), but I see how...

Agreed on both points. I think the separate file makes sense if we feel like the algorithm to select the best file needs to get more complicated.

I looks forward to hearing what you think.

---
++ With no other identifier use, you might think that the namespace is unused (and maybe we should report that). It's the case for a namespace alias, but not a namespace directive. Consider:

```
headerA.h:
namespace A {
    void functionA();
}

headerB.h:
namespace B {
   using namespace A;
   void functionB();
}

fileC.cc:
namespace C {
  // is this used?
  using namespace B;

  void myFunc() {
     // namespace B pulls in namespace A, so I can call functionA()!
     // So the using namespace B above is 'used', but just by considering
     // the decl's in use, you will only detect namespace A.
     functionA();
  }
}
```

I had been looking at whether I could detect this use of the namespace directive, but it involves detecting when the modified namespace search path has an affect on the symbol lookup.

There's further issues if we push that example further though. What if the `using namespace A` only occurs in a specific header, and that's not the one we suggest. My commit could be modified to further prioritise a `NamespaceDecl` containing using directives.

I think this example is degenerate, and you shouldn't be relying on the `using namespace B` to get a symbol from `A` - but I guess it's possible for a library to do this to re-export an internal namespace.

Apologies this comment ended up much longer than I planned...

bolshakov-a commented:
The first commit LGTM, at least conceptually.

Regarding to the second commit, I agree that it is annoying when IWYU suggests to include a header for the stuff (using-directive, e. g.) which you just forgot to remove. But, maybe, it should be concern of some other tool?

As for the fix of the original issue (the 3rd commit), honestly, it seems redundant to me to add some complex logic just for attribution to the most appropriated header in explanatory comments. But I like the idea that it should be in the pending analysis. Maybe, just make it more similar to `using_decl_referenced_` handling, like you've done for namespace aliases?
> What if the using namespace A only occurs in a specific header, and that's not the one we suggest.

Good catch! I suspect it may be a tricky issue which may be addressed some day later. But, probably, your tracking of declarations inside namespaces actually forms a good framework to this?

Concerning my commit, feel free to cherry-pick it or just to take inspiration from it, if you wish.

kimgr commented on iwyu.cc:
```diff
@@ -1699,6 +1700,17 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       ReportTypeUse(used_loc, type);
   }
 
+  void ReportNamespaceAliasUse(SourceLocation used_loc,
```
Can we skip the `Track unused namespace aliases` commit for now? I'm not sure if it's trying to do some additional cleanup of unused aliases or if it's a foundation for IWYU analysis. If the former, I don't think IWYU should do it. If the latter, it makes sense to tackle separately, I think.

kilroyd replied:
Sure. This isn't a foundation for IWYU analysis.

Having played with the namespace reporting, I feel that as long as IWYU suggests a header to satisfy a reference to a namespace, then in order minimise unused headers IWYU also needs to identify unused aliases and using directives. I've got had some ideas on how we might be able to do this better.

kimgr commented on iwyu.cc:
```diff
@@ -2578,6 +2578,12 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   // Visitors defined by BaseAstVisitor.
 
   bool VisitNestedNameSpecifier(NestedNameSpecifier* nns) {
+    // If this is a NamespaceAlias, then mark the alias as used,
```
Hmm, out of curiosity, why isn't the check for `NestedNameSpecifier::Namespace` necessary here too?

kilroyd replied:
I'm afraid you've lost me. This hunk is about detecting when an identifier is preceded by an explicit namespace alias, in which case we want to report the use of the alias.

As far as I understand it, we can't detect using-directive usage here (at least not in a way that tracks usage completely).

kimgr replied:
Right, so:
```
using alias = std;

void foo() {
    alias::vector<int> x;  // hits your VisitNestedNameSpecifier check
    std::vector<int> x;  // why don't we want to check this in VisitNestedNameSpecifier too?
}
```

kilroyd replied:
Note that this hunk will be removed once we drop commit 2 i.e. warning on unused aliases.

In VisitNestedNameSpecifier() we're able to identify alias use (and therefore spot unused aliases).

We could consider std::vector too, and try to identify potential using-directives that are being used. But this won't be able to catch all uses appropriately. std::vector specifically is generally not 'using' any using-directive. As I think I've mentioned previously, identifying where a using-directive is used requires hooking into the symbol lookup infrastructure of clang (Sema/SemaLookup.cpp).

kilroyd replied:
> Note that this hunk will be removed once we drop commit 2

I got confused between the commits - this was in the first commit so still being added.

kimgr replied:
Yeah, I'm not saying we should do anything, I was just curious why alias uses were different from other nested name-specifiers. Maybe because a namespace nns ultimately refers to a symbol that is guaranteed to live in the same file as the namespace. The alias, on the other hand, could have been included from somewhere that has nothing to do with the aliased namespace/symbol. OK, that makes sense to me, whether it's correct or not :)

kimgr commented on iwyu_namespace.cc:
```diff
@@ -0,0 +1,249 @@
+    // as the function/structure will be.
+    //
+    // TODO: others?
+    if (const clang::FunctionDecl* func = DynCastFrom(decl_context)) {
```
Are these boundary conditions necessary? In the majority of cases, I think there might be a `RecordDecl` at most between these and the `TranslationUnitDecl`, so if it's an optimization to exit early, I don't think it's going to buy us very much.

Or does it have any functional impact?

kilroyd replied:
This has a functional impact on the count of declarations made within a namespace. `RecordNamespaceDecls()` is being called for at least function arguments and structure fields. I believe it's getting called for variable declarations within function scope too.

So these aborts ensure that a function within a namespace only results in an increment of 1 (on all parent namespaces), similarly a struct.

Things _may_ work ok with these removed, but it's easier to validate the counts in the error log if they conform to what you'd expect.

kimgr replied:
Side note: you can replace the three `DynCastFrom` with `isa<FunctionDecl, TagDecl, BlockDecl>(decl_context)`

kilroyd replied:
Yes, that's right.

kimgr commented on iwyu_namespace.cc:
```diff
@@ -0,0 +1,249 @@
+
+  // Ignore declarations of sub-namespaces i.e. don't count them as a
+  // forward decl or full decl.
+  if ((ns = DynCastFrom(decl)))
```
Prefer `isa<NamespaceDecl>(decl)` here

kimgr reviewed (commented):
I've turned this inside out a few times, and I think I roughly see what's going on. It looks like you're maintaining a structure like this:
```
{
    qualified-name: {
        filename: fileinfo(decl, full_count, fwd_count),
        filename: fileinfo(decl, full_count, fwd_count),
        filename: fileinfo(decl, full_count, fwd_count),
    },
    qualified-name: {
        filename: fileinfo(decl, full_count, fwd_count),
        filename: fileinfo(decl, full_count, fwd_count),
        filename: fileinfo(decl, full_count, fwd_count),
    }
}
```

I'm pretty sure you don't need the qualified-name in the mix, and I'm not entirely sure you need the keying by filename. I would suggest something like:
```
{
    canonical_decl: {decl: count},
    canonical_decl: {decl: count},
    canonical_decl: {decl: count},
}
```
instead.

* `NamespaceDecl::getCanonicalDecl()` should return a stable identity pointer for a namespace declaration chain. So for any `NamespaceDecl` you can call `getCanonicalDecl` and get something that identifies the namespace just like its qualified-name. I think.
* The forward-decl count is never used, so I suggest we get rid of it
* I don't think the filepath component of the index is necessary (happy to be proven wrong!), so you should be able to just use `GetFilePath(namespace_decl)` for diagnostics and whatnot.
* The `nsmap` appears to be a cache for `GetBestNamespaceDecl`. Let's skip that for now, unless performance suffers unbearably.
* I suspect those changes are going to collapse enough that the namespace use-count map could hang off `IwyuFileInfo` just like the `UsingDecl` used map. As far as I can tell it's not absolutely necessary that the namespace map is global. There might be some redundancy between different files, but maybe it also gives more granular results if broken down by use-file?
* Some style nits here and there, but we can discuss those once the bigger things have been ironed out.

Thanks!

kilroyd reviewed (commented):
> It looks like you're maintaining a structure like this:

Correct.

> I'm pretty sure you don't need the qualified-name in the mix, and I'm not entirely sure you need the keying by filename. I would suggest something like ... instead

I'm not quite sure what you're proposing here... I'll need to have a think - it's been a couple weeks since I wrote it, but my initial reaction is that structure isn't sufficient.

> NamespaceDecl::getCanonicalDecl() should return a stable identity pointer for a namespace declaration chain. So for any NamespaceDecl you can call getCanonicalDecl and get something that identifies the namespace just like its qualified-name

Yes, `getCanonicalDecl()` can be used to identify the namespace like qualified-name. So we could swap the qualified-name index for canonical-decl. Other than that, the canonical-decl isn't useful.

We still need to record declaration counts separately for each file during ?ast travesal? so we can decide which is best in post-traversal analysis.

I'll see what I can do to address your comments...

kimgr commented:
> Yes, getCanonicalDecl() can be used to identify the namespace like qualified-name. So we could swap the qualified-name index > for canonical-decl. Other than that, the canonical-decl isn't useful.

Right; but it's much cheaper to carry and compare than a string.

> We still need to record declaration counts separately for each file during ?ast travesal? so we can decide which is best in
> post-traversal analysis.

You could record declaration counts for each _decl_ during AST traversal, and then aggregate for files post-traversal. So during traversal we could pick up:
```
canonical_decl: [(decl, count), (decl, count), (decl, count)]
```
all of which are register-sized and cheap to get. Then before analysis, you could transform the resulting set into a per-file breakdown, possibly precomputing the best decl for every canonical decl.

Lookup is then just `best_decls[ns->getCanonicalDecl()]`. Feels like that should be doable, but I haven't gotten around to trying it.

kilroyd commented:
I've incorporated most of the changes from the kg-namespaces branch. Thanks for the heap cleanup - my background is C, so I fall back to using the heap when I can't figure out the C++ way to do things :)

Also made tweaks to the algorithm to track using directives, and prefer namespaces with them.

Not done:
* Still have filepath key in the map - I think we need to keep this, so we count the declarations within a file across multiple namespace blocks (with the same qualified name). I've expanded comments to hopefully explain better
* I dropped your commit with early out on forward declaration in RecordDeclNamespaces(). I had tried this independently too. I think we can do this, but at the moment I'm relying on the verbose output printing forward decl info. It has the effect of not creating an IwyuNsFileInfo which has a subtle effect on the reporting in getBestFile(). I expect to take another look. 

Still todo:
* Consider merging IwyuNsFileInfo into IwyuFileInfo. Global ns info vs per file.
* Double check behaviour vs Andrey's commit
* testcases for:
    * prefer namespace with using directives
    *  ??

kimgr commented on iwyu_namespace.h:
```diff
@@ -0,0 +1,26 @@
+}  // namespace clang
+
+namespace include_what_you_use {
+void RecordDeclInNamespaces(const clang::NamedDecl *decl);
+
+const clang::NamespaceDecl *GetBestNamespaceDecl(
+    const clang::NamespaceDecl *ns);
```
We bind the pointer asterisk to the type in IWYU (because Google conventions). Clang-format should have fixed that.

kimgr replied:
Turns out our `.clang-format` didn't configure this correctly, I pushed a fix to master, so if you format this again, it should align the stars correctly.

kimgr commented on iwyu_namespace.h:
```diff
@@ -0,0 +1,26 @@
+void RecordDeclInNamespaces(const clang::NamedDecl *decl);
+
+const clang::NamespaceDecl *GetBestNamespaceDecl(
+    const clang::NamespaceDecl *ns);
```
I would remove the parameter name `ns` here, because that would make the whole declaration fit on one 80-col line.

kimgr commented on iwyu_output.cc:
```diff
@@ -721,6 +722,15 @@ void IwyuFileInfo::ReportUsingDeclUse(SourceLocation use_loc,
+void IwyuFileInfo::ReportNamespaceDeclUse(SourceLocation use_loc,
+                                          const NamespaceDecl* decl,
+                                          UseFlags flags, const char* comment) {
+  // In order to assign Namespace usage to the best file, don't record the
```
"Namespace" -> "namespace"

kimgr commented on iwyu_output.cc:
```diff
@@ -2204,6 +2214,14 @@ void IwyuFileInfo::ResolvePendingAnalysis() {
+  for (NamespaceDeclUse& use : namespace_uses) {
+    const NamespaceDecl* decl = GetBestNamespaceDecl(use.decl);
+    if (decl != nullptr) {
+      // For each namespace use, report against the best NamespaceDecl
```
I'd move this comment to just before the loop. Also, add a full stop to end the sentence.

kimgr commented on iwyu_output.cc:
```diff
@@ -2204,6 +2214,14 @@ void IwyuFileInfo::ResolvePendingAnalysis() {
     }
   }
+
+  for (NamespaceDeclUse& use : namespace_uses) {
```
Can be `const`

kimgr commented on iwyu_output.h:
```diff
@@ -216,6 +216,13 @@ class OneIncludeOrForwardDeclareLine {
   const clang::NamedDecl* fwd_decl_;
 };
 
+struct NamespaceDeclUse {
```
Could you add a little doc comment to this type? Just something around how we delay uses of namespace decls and finalize them in ResolvePendingAnalysis.

kimgr commented on iwyu_output.h:
```diff
@@ -297,6 +304,13 @@ class IwyuFileInfo {
 
+  // Called whenever a NamespaceDecl is used, generally via a using
+  // directive or an alias (a use via a symbol will force the right
+  // include via that symbol)
```
Full stop at end of comment

kimgr commented on tests/cxx/multiple_decl_namespace-d1.h:
```diff
@@ -0,0 +1,21 @@
+const Flags F_None = 0;
+const Flags F_All = 1;
+
+void function1(const test::ns3::Class3 &c, Flags f);
```
Let's use IWYU naming conventions in tests too; functions are `PascalCase`. Here and throughout the new tests

kimgr commented on tests/cxx/multiple_decl_namespace.cc:
```diff
@@ -0,0 +1,55 @@
+// Tests that we handle multiply declared namespaces correctly.
+//
+// Where declarations of a namespace occur in multiple headers (either
+// because forward declarations were needed or otherwise), iwyu should
```
I try to spell `iwyu` as `IWYU` in all new code/docs, so could you stick to that here and below?

kimgr commented on tests/cxx/multiple_decl_namespace.cc:
```diff
@@ -0,0 +1,55 @@
+
+/**** IWYU_SUMMARY
+
+(tests/cxx/multiple_decl_namespace.cc has correct #includes/fwd-decls)
```
I try to avoid succeeding testcases like this, because a lot of subtlety and information is lost. See https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/IWYUStyle.md#force-diagnostics.

See if you can move the symbols out of `-dx.h` headers and into `-ix.h` headers so IWYU is forced to complain about the indirect includes.

kimgr commented on iwyu_output.h:
```diff
@@ -216,6 +216,13 @@ class OneIncludeOrForwardDeclareLine {
+  clang::SourceLocation used_loc;
+  const clang::NamespaceDecl* decl;
+  UseFlags flags;
+  const char* comment;
```
I wonder if this needs to become a `std::string` to avoid lifetime issues. We mostly call `ReportXXXUse` with literal strings, but should someone add something like a concatenated string and use `c_str()`, this would refer to a dead value.

In the current flow, that doesn't matter, because the comment will be copied into `OneUse`, and materialized in a new `stdd:string`, but this intermediate storage is a little risky.

kilroyd replied:
Good spot. I'll push an update fixing this and the majority of the issues you've raised. I still need to have a go at moving the logic to `IwyuFileInfo` which I'm struggling to find the time for. Maybe later this weekend🤞?

kimgr replied:
Thanks! I've been a little occupied with other things as well, so definitely no rush from my side. Let's percolate on it until it feels stable and nice.

headhunter45 commented:
I'm probably misunderstanding this, but if I have ```#include <string>
using namespace std;``` at the top of a file and I don't actually use std::string or anything else from the std namespace what will this do? I'd want it to remove both the include and the using statement, but I think having the using would keep the `#include <string>` in this case. and IWYU wouldn't remove the using because I guess that's out of scope for the tool?

kimgr commented:
@headhunter45 Yes, exactly. Otherwise it's hard to know when to stop - if you have an unused static function in your file that uses `std::string`, do we remove that? Etc.

IWYU's focus is the relationship between symbols and files, and I think it's nice to stay within those fences.

That said, it would be cool to see some clang-tidy checks to do e.g. unused `using` cleanup. Then IWYU would be free to remove the leftover headers.

kimgr commented on iwyu_output.h:
```diff
@@ -320,6 +368,10 @@ class IwyuFileInfo {
     return direct_includes_as_fileentries_;
   }
 
+  void RecordDeclInNamespaces(const clang::NamedDecl* decl);
```
I understand that `RecordDecl` is a verb phrase here ("record the declaration"), but `RecordDecl` is also a type in Clang ("a record declaration"). Would be nice to find a different word than `Record` :-)

kimgr commented on iwyu_output.h:
```diff
@@ -417,6 +477,30 @@ class IwyuFileInfo {
+  //     },
+  //   etc
+  // }
+  map<const clang::NamespaceDecl*, map<const IwyuFileInfo*, IwyuNamespaceInfo>>
```
This can be a `FileEntry*` for simpler implementation.

kimgr commented on iwyu.cc:
```diff
@@ -1700,8 +1701,25 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
   // Visitors of types derived from clang::Decl.
+  bool VisitNamedDecl(clang::NamedDecl* decl) {
+    preprocessor_info()
+        .FileInfoFor(preprocessor_info().main_file())
```
This use of `main_file()` was surprising to me. I tried to replace it with `CurrentFileEntry()`, but that caused test failues.

I think I understand now that this is why the namespace info was global before? Because it's statistics over the whole compilation unit, not really keyed by file?

kilroyd replied:
To do this the statistics must be recorded per namespace. Under that namespace we subkey by file, allowing us to identify the best file for the namespace.

If we reverse the keys, recording stats on each namespace declared within a file, then we still need something to tell us which files a given namespace is in, or we have to iterate over all files for each namespace we want to find.

So this does look a bit odd, with the `nsinfoPerFile` map is in the `IwyuFileInfo`. If processing multiple translation units in one run, this should make the namespace processing independent for those translation units, which is a good thing.

kimgr replied:
Sorry, I'm not following. You keep saying "this" about something that sounds interesting, but I'm not sure what it is :)

But yeah, the way you put all statistics about namespaces into the main file make me think that moving the state to `IwyuFileInfo` was not the right move.

> If processing multiple translation units in one run, this should make the namespace processing independent
> for those translation units, which is a good thing (not sure if IWYU actually does that).

IWYU only does one translation unit at a time.

But even if it did multiple, I think putting this state into the AST visitor might be a better fit. Let me play around with a refactor for the UsingDecl delayed reporting, I think there may be symmetries/synergies...

kilroyd replied:
> Sorry, I'm not following. You keep saying "this" about something that sounds interesting, but I'm not sure what it is :)

My apologies, I've assumed we're on the same page. I guess this PR has been open a while.

> To do this the statistics...

Here I was refering to the PR, 'this' being "finding the best include file for a given namespace".

> So this does look a bit odd...

This instance refers to the current commit. :)

I'll wait for further feedback. I can back up to a version with all the changes except the move into `IwyuFileInfo`. Just let me know your preference.

kimgr replied:
> Let me play around with a refactor for the UsingDecl delayed reporting, I think there may be symmetries/synergies...

Nope, the UsingDecl handling seems slightly different. It seems to keep track of uses within the same file. And it's not remembering uses to replay them later, it has enough information in the UsingDecl to create the uses from scratch. Or at least that's what it does :). Nevermind.

kilroyd replied:
FYI, I'm reworking this commit so the `IwyuNamespaceInfo` is stored in `IwyuFileInfo` lists as you might expect. The `nsPerFile` map be static and point to these entries, allowing us to do the lookup during `GetBestNamespaceDecl` as we need to.

kimgr replied:
Sorry for the prolonged silence, I've been alternating between busy with work and vacationing. I'm actually not sure anymore that the namespace statistics should go in `IwyuFileInfo`, given how they're global across files.

I'm a little cautious around static members, because they tend to cause trouble with reentrancy (and threading, which is not an immediate concern here), but let me take a look at your latest patch and see if I can get my head around it.

kimgr commented:
It hadn't occurred to me that the two commits are pretty much unrelated, right? Maybe it would make sense to split this PR up into two with one commit each. It's hard to review these large PRs, particularly with many new tests.

kimgr commented on iwyu.cc:
```diff
@@ -183,6 +183,7 @@ using clang::LValueReferenceType;
 using clang::LinkageSpecDecl;
```
Commit message: please use "IWYU" over "iwyu" consistently.

kimgr commented on iwyu_output.h:
```diff
@@ -222,6 +224,45 @@ class OneIncludeOrForwardDeclareLine {
+// Namespace usage reporting is delayed until ResolvePendingAnalysis.
+// This structure stores the arguments to be passed to
+// RecordFullSymbolUse.
+struct NamespaceDeclUse {
```
I think we can just use `OneUse` to buffer pending namespaces uses, instead of this custom type

kilroyd replied:
You're right - It does hold much of the same data, but has quite a bit more too. I could see this being confusing. Maybe a tuple might be prefered here?

kilroyd commented:
> It hadn't occurred to me that the two commits are pretty much unrelated, right?

TBH I'd forgotten there were 2 commits. They're both modifying reporting on namespace usage. If we commit the reporting on namespace aliases without improving the suggested filenames, then we end up with more inappropriate headers being suggested for a given project. If you want them separate, I'll drop the namespace alias commit.

kimgr commented:
@kilroyd I think that's fine; they solve independent problems that are both increments toward better behavior.

I have some review feedback on the first (namespace alias, right?) commit, but that's almost 100% ready to be merged, so it seems like a shame to block it on the the second (use-stats-based namespace attribution), which is more complex.

Could you pull the namespace alias commit into its own PR? I'd also be happy to adopt it and do the grunt work if you prefer.

kimgr commented:
I have some spare time for opensource work now, so if you don't mind, I'll play around with this patch set and experiment a little. I think that's more likely to make progress than bouncing review comments back and forth. Won't merge anything without running it by you first, we'll see how far I get :)

kimgr commented:
I've pulled on various strings and have something running now where:

* A new class `IwyuNamespaceStats` is added to `iwyu_output.{h,cc}`
* The class encapsulates the two-level map and exposes an `Update` method (previously `ReportDeclInNamespace`) and a `GetBestNamespaceDecl`, as before
* An `IwyuNamespaceStats` is held in the visitor state in `iwyu.cc`
* A reference to that stats object is passed to `ResolvePendingAnalysis`
* That way no globals are necessary, and `iwyu.cc` is mostly responsible for accounting namespace stats, whereas `iwyu_output` is responsible for reporting namespace uses based on those stats
* I can share the branch when I've cleaned it up a little

However, now I'm looking at the logic for `GetBestNamespaceDecl`, and I'm not sure I understand the rationale... For example;

* the decl count does not appear to be used
* I'm trying to get my head around this comment in the test:
```
// This namespace has a using-directive. This is a pretty clear sign
// that this namespace block is not a bunch of forward declarations,
// and need to be explicitly included by users.
```
-- why is that a clear sign?
* I was expecting `GetBest...` to do some kind of scoring, but it really seems to start arbitrarily with the first file, and then focusing primarily on the number of using-directives in the namespaces. Is it possible to clarify a little around the rules for forward-declarations, using-directives and multiple namespaces with usings and decls?

kilroyd commented:
>     * the decl count does not appear to be used
> 
>     * I'm trying to get my head around this comment in the test:
> 
> 
> ```
> // This namespace has a using-directive. This is a pretty clear sign
> // that this namespace block is not a bunch of forward declarations,
> // and need to be explicitly included by users.
> ```
> 
> -- why is that a clear sign?
> 
>     * I was expecting `GetBest...` to do some kind of scoring, but it really seems to start arbitrarily with the first file, and then focusing primarily on the number of using-directives in the namespaces. Is it possible to clarify a little around the rules for forward-declarations, using-directives and multiple namespaces with usings and decls?

I had to double check what we have in `GetBestNamespaceDecl`. The function encapsulates the decision logic, and we should be able to modify it to do whatever we want. I've tried to keep the logic as simple as possible, and from the look of it, avoiding incorrect suggestions.

Currently the logic is:
1. Only files declaring a namespace that has full declarations or using directives are candidates.
2. A file declaring a namespace containing a using-directive is chosen over a namespace without.
3. If more than one file declares a namespace with a using-directive, we don't try to pick one.
4. If more than one file declares a namespace with full declarations, we don't try to pick one.

A namespace with only forward declarations should be dropped at 1.

Next, consider two coding styles:
* single header containing all the interfaces within a given namespace
* multiple headers defining the content of a namespace, each defining a different aspect.

We can't handle the multiple header case without tracking which declarations we are actually using from each namespace declaration. This is equivalent to 4. (and probably 3.) - we avoid picking a header. This is why the full declaration count is not used in choosing a header. It would result in always choosing the file with the 'most declarations' or widest interface, but that may not be what is being used.

A using-directive within a namespace modifies the symbol lookup within that namespace.

```namespace a::b::c::d {
struct m;
}

namespace a::e {
  using namespace a::b::c::d;
  // symbol m is now available in the namespace a::e (as is everything else in a::b::c::d)
}

// Not limited to the scope
a::e::m x;

namespace a::e {
   // and still available in later scope
   m y;
}
```

https://godbolt.org/z/1aWsnv6W1

The persistence of the effect of the using-directive outside the namespace scope significantly changes how you would refer to `a::b::c::d` between picking up a namespace with the using-declaration vs without. This is essentially why I added the prioritisation of using-directives.

Thinking about this again, if this were an internal namespace that was being re-exposed then you would never have used `a::b::c::d` anyway... and it's possible that you only need to `a::b::c::d` with this partial interface. Maybe we should drop the using-directive check...

kimgr commented:
> We can't handle the multiple header case without tracking which declarations we are actually using from each namespace declaration.

So I guess I'm wondering: why can't we do this? Just intuitively, it seems more relevant to look at usage statistics than definition statistics, but I don't know if there's any concrete pros/cons...

kilroyd commented:
This has been dragging on a bit, and I haven't had time to address the issues. I propose dropping this RFC, and instead will push a commit that mostly ignores namespace declarations, along the lines of the @bolshakov-a 's commit. That should be sufficient for my interests, is probably good enough for 90% of cases, and is much simpler.

kimgr commented:
@kilroyd Alright, thanks for investing so much energy this far. I think there's some gold in this approach, but it's not 100% clear to me how to isolate the "best" namespace. Intuitively I don't think number of declarations is it, and I'm not sure if number of uses is much better. I'll see if I can muster time and energy to get back to this later.

kilroyd commented:
I'll leave the namespaces branch in my repo in case it's useful to resurrect in future
